module.nix

```diff
@@ -57,23 +64,29 @@ in
 
         User = "harmonia";
         Group = "harmonia";
+        DynamicUser = true;
 
         RuntimeDirectory = "harmonia";
+        LoadCredential = lib.optional (cfg.signKeyPath != null) "sign-key:${cfg.signKeyPath}";
 
-        SystemCallFilter = "@system-service";
+        SystemCallFilter = [
+          "@system-service"
+          "~@privileged"
+          "~@resources"
+        ];
 
         PrivateNetwork = false;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        ProtectProc = true;
+        NoNewPrivileges = true;
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6";
 
         LimitNOFILE = 65536;
       };
     };
-
-    users = {
-      users.harmonia = {
-        isSystemUser = true;
-        group = "harmonia";
-      };
-      groups.harmonia = { };
-    };
```

Comment:
This broke some logic I inherited from @NinjaTrappeur that was passing a signing key into Harmonia from a sops location:

```
      nix-signing-key = {
        restartUnits = [ "harmonia.service" ];
        owner = config.users.users.harmonia.name;
        mode = "0440";
      };
```

Looks like the two ways forward here are systemd's `LoadCredential` scheme (which will make a file available read-only to the unit's user/environment) or else switching from a file to an environment variable. Any thoughts on which of these would be preferable or if there's another option I'm missing?

Reply:
Ah, there's an existing discussion on this in https://github.com/Mic92/sops-nix/issues/198.

Reply:
LoadCredential that is used, shouldn't require a user. Here an example:

https://github.com/Mic92/dotfiles/blob/e8829eed755bf154579668033c962c962ef6a943/nixos/eve/modules/harmonia.nix#L3

Reply:
Thanks! Yes, that's helpful and I do now see how the pieces fit together. I think our original deployment predated the introduction of LoadCredential on your side.